File: Android/ui/src/main/java/com/felipearpa/tyche/ui/lazy/FirstTimeLoadingVisibilityDecider.kt
package com.felipearpa.tyche.ui.lazy

import androidx.paging.LoadState
import androidx.paging.compose.LazyPagingItems

class FirstTimeLoadingVisibilityDecider<Value : Any> : LoadingVisibilityDecider<Value> {
    var hasShownLoading = false
    var isContentLoaded = false

    override fun shouldShowLoader(lazyPagingItems: LazyPagingItems<Value>): Boolean {
        if (lazyPagingItems.isNotEmpty) {
            return false
        }

        if (lazyPagingItems.loadState.refresh is LoadState.Loading) {
            hasShownLoading = true
        }

        when (lazyPagingItems.loadState.refresh) {
            is LoadState.NotLoading, is LoadState.Error ->
                if (hasShownLoading) {
                    isContentLoaded = true
                }

            else -> {}
        }

        return lazyPagingItems.loadState.refresh is LoadState.Loading && !isContentLoaded
    }
}

File: iOS/UI/Sources/UI/Paging/FirstTimeLoadingVisibilityDecider.swift
import Foundation

public class FirstTimeLoadingVisibilityDecider<Key, Value>: LoadingVisibilityDecider {
    private var hasShownLoading = false
    private var isContentLoaded = false

    public init() {}

    @MainActor
    public func shouldShowLoader(lazyPagingItems: LazyPagingItems<Key, Value>) -> Bool {
        if lazyPagingItems.isNotEmpty {
            return false
        }

        if lazyPagingItems.loadState.refresh.isLoading {
            hasShownLoading = true
        }

        switch lazyPagingItems.loadState.refresh {
        case .notLoading, .failure:
            if hasShownLoading {
                isContentLoaded = true
            }
        default:
            break
        }

        return lazyPagingItems.loadState.refresh.isLoading && !isContentLoaded
    }
}
